gcc/cp/contracts.cc

```diff
@@ -3464,150 +3394,245 @@ p2900_duplicate_contracts (tree newdecl, tree olddecl)
       cont_end = make_location (new_loc, new_loc, cont_end);
       error_at (cont_end, "declaration adds contracts to %q#D", olddecl);
       inform (rd.note_loc , "first declared here");
-      /* We have stripped the contracts from the new decl, so that they will
-	 not be merged into the original decl (which had none).  */
+
       return;
     }
 
   if (old_contracts && !new_contracts)
-    /* We allow re-declarations to omit contracts declared on the initial decl.
+    {
+      /* We allow re-declarations to omit contracts declared on the initial decl.
        In fact, this is required if the conditions contain lambdas.  Check if
        all the parameters are correctly const qualified. */
-    check_param_in_redecl (olddecl, newdecl);
-  else if (contract_any_deferred_p (new_contracts)
-	   || contract_any_deferred_p (old_contracts))
-    /* TODO: stash these and figure out how to process them later.  */
-    ;
+      check_param_in_redecl (olddecl, newdecl);
+    }
+  else if (old_contracts && new_contracts &&
+      !contract_any_deferred_p (old_contracts)
+      && contract_any_deferred_p (new_contracts)
+      && DECL_UNIQUE_FRIEND_P (newdecl))
+    {
+	/* Newdecl's contracts are still DEFERRED_PARSE, and we're about to
+	   collapse it into olddecl, so stash away olddecl's contracts for
+	   later comparison.  */
+	defer_guarded_contract_match (olddecl, olddecl, old_contracts);
+	/* put the defered contracts on the olddecl so we parse it when
+	  we can.  */
+	copy_deferred_contracts(newdecl, olddecl);
+    }
+  else if (contract_any_deferred_p (old_contracts)
+	   || contract_any_deferred_p (new_contracts))
+    {
+      /* TODO: ignore these and figure out how to process them later.  */
+      /* Note that an in class friend declaration has deferred contracts, but out
```

Comment:
I think we should talk about "a friend defined in a class definition" and "a friend declared but not defined in a class definition" here, "out of class friend declaration" doesn't sound like a thing that exists, and all friend declarations are "in class".